File: kernel-r/declarativewidgets/R/widget_dataframe.r
#' @include widget.r serializer.r querier.r
NULL
#' Widget_Dataframe
#'
#' Description
Widget_Dataframe <- R6Class(
    'Widget_Dataframe',
    inherit = Widget,
    public = list(
        serializer = NULL,
        limit = NULL,
        query = NULL,
        variable_name = NULL,
        querier = NULL,
        handle_backbone = function(msg) {
            msg_limit <- msg$sync_data$limit
            msg_name <- msg$sync_data$variable_name
            msg_query <- msg$sync_data$query
            if(!is.null(msg_limit)) {
                self$register_limit(msg_limit)
            } else {
                log_info("No limit value provided for Widget_Dataframe")
            }
            if(!is.null(msg_name)) {
                self$register_name(msg_name)
            } else {
                log_info("No name value provided for Widget_Dataframe")
            }
            if(!is.null(msg_query)) {
                self$register_query(msg_query)
            } else {
                log_info("No name query provided for Widget_Dataframe")
            }
        },
        handle_custom = function(msg) {
            if(!is.null(msg$event)) {
                if(msg$event == 'sync') {
                    self$serialize_and_send(self$variable_name, self$limit, self$query)
                } else {
                    log_info(paste("Unhandled custome event:",msg$event))
                }
            } else {
                log_info("No event value in custom Widget_Dataframe comm message")
            }
        },
        df_in_interpreter = function(name) {
            tryCatch({
                eval(name)
            }, error = function(e) {
                return (FALSE)
            })
            return (TRUE)
        },
        get_df_in_interpreter = function(name) {
            df_name_vector <- unlist(strsplit(name, "$", fixed=TRUE))
            if(length(df_name_vector) == 2) {
                var_type <- class(get(df_name_vector[[1]], envir = .GlobalEnv))[[2]]
                if(var_type == "R6") {
                    return (get(df_name_vector[[1]], envir = .GlobalEnv)[[df_name_vector[[2]]]])
                } else {
                    log_info(paste("The df class type", var_type, "is not supported"))
                    return (NULL)
                }
            } else {
                return (get(name, envir = .GlobalEnv))
            }
        },
        serialize_and_send = function(name, limit, query = list()) {
            if(self$df_in_interpreter(name)) {
                #apply query before sending over
                df_after_query <- self$querier$apply_query(self$get_df_in_interpreter(name), query)
                serialized_df <- self$serializer$serialize(df_after_query, limit)
                self$send_update("value", serialized_df)
                return (TRUE)
            } else {
                log_info(paste("DataFrame ", name, " not found! No sync message sent."))
                return (paste("Dataframe", name, "not found!"))
            }
        },
        register_name = function(name) {
            self$variable_name <- name
            response <- self$serialize_and_send(name, self$limit)
            self$handle_function_response(response)
        },
        register_limit = function(limit) {
            self$limit <- limit
        },
        register_query = function(query) {
            self$query <- fromJSON(query)
            response <- self$serialize_and_send(name, self$limit, self$query)
            self$handle_function_response(response)
        },
        initialize = function(comm, serializer, querier) {
            #initialize super class Widget and serializer
            super$initialize(comm)
            self$serializer <- serializer
            self$querier <- querier
        }
    )
)

File: kernel-scala/src/main/scala/declarativewidgets/WidgetDataFrame.scala
/**
 * Copyright (c) Jupyter Development Team.
 * Distributed under the terms of the Modified BSD License.
 */

package declarativewidgets

import declarativewidgets.query.QuerySupport
import declarativewidgets.util.{MessageSupport, SerializationSupport}
import org.apache.spark.sql.DataFrame
import org.apache.toree.comm.CommWriter
import org.apache.toree.interpreter.Interpreter
import org.apache.toree.kernel.protocol.v5.MsgData
import play.api.libs.json._

/**
 * A widget for retrieving the value of a DataFrame in the kernel.
 *
 * @param comm CommWriter used for communication with the front-end.
 */
class WidgetDataFrame(comm: CommWriter)
  extends Widget(comm) with SerializationSupport with QuerySupport{

  // name of the DataFrame in the kernel
  var variableName: String = _

  // maximum number of rows to send
  var limit: Int = Default.Limit

  // the query to apply to the DataFrame. Defaults to JSON empty array.
  var query: String = "[]"

  lazy val kernelInterpreter: Interpreter = getKernel.interpreter

  /**
   * Handles a Comm Message whose method is backbone.
   * @param msg The Comm Message.
   */
  override def handleBackbone(msg: MsgData, msgSupport:MessageSupport): Unit = {
    logger.debug(s"Handling backbone message ${msg}...")
    (msg \ Comm.KeySyncData \ Comm.KeyLimit).asOpt[Int].foreach(registerLimit(_))

    (msg \ Comm.KeySyncData \ Comm.KeyDataFrameName).asOpt[String].foreach(registerName(_))

    (msg \ Comm.KeySyncData \ "query").asOpt[String].foreach(registerQuery(_))
  }

  /**
   * Handles a Comm Message whose method is custom.
   * @param msgContent The content field of the Comm Message.
   */
  override def handleCustom(msgContent: MsgData, msgSupport:MessageSupport): Unit = {
    logger.debug(s"Handling custom message ${msgContent}...")
    (msgContent \ Comm.KeyEvent).asOpt[String] match {
      case Some(Comm.EventSync) =>
        syncData(msgSupport)
      case Some(evt) =>
        logger.warn(s"Unhandled custom event ${evt}.")
      case None =>
        logger.warn("No event value in custom comm message.")
    }
  }

  private[declarativewidgets] def registerLimit(limit: Int): Unit = {
    this.limit = limit
    logger.debug(s"Registered limit ${limit}.")

  }

  private[declarativewidgets] def registerName(name: String): Unit = {
    this.variableName = name
    logger.debug(s"Registered DataFrame variable name ${name}.")
  }

  private[declarativewidgets] def registerQuery(query: String): Unit = {
    this.query = query
    logger.debug(s"Registered query ${query}.")
  }

  private[declarativewidgets] def theDataframe = {
    kernelInterpreter.read(variableName)
  }

  private[declarativewidgets] def syncData(msgSupport:MessageSupport) = {

    val result: Either[String, JsValue] = theDataframe match {
      case Some(df:DataFrame) =>
         try{
           val query = Json.parse(this.query).asOpt[JsArray].getOrElse(new JsArray())
           Right(serialize(applyQuery(df, query), this.limit))
         }
        catch {
          case parseError:com.fasterxml.jackson.core.JsonParseException =>
            parseError.printStackTrace()
            Left(parseError.getMessage)
        }

      case Some(_) =>
        logger.warn(s"${this.variableName} is not a DataFrame")
        Left(s"${this.variableName} is not a DataFrame")

      case None =>
        logger.warn(s"DataFrame ${this.variableName} not found! No sync message sent.")
        Left(s"DataFrame ${this.variableName} not found!")
    }

    result match {
      case Right(serDf) =>
        sendSyncData(serDf, msgSupport)
        msgSupport.sendOk()

      case Left(error) =>
        msgSupport.sendError(error)
    }
  }

  private[declarativewidgets] def sendSyncData(result: JsValue,msgSupport:MessageSupport) =
    msgSupport.sendState(Comm.StateValue, result)

}
